src/lesson_4/FractionalWrapper.sol

```diff
@@ -0,0 +1,176 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.12;
+
+import {ERC20} from "yield-utils-v2/contracts/token/ERC20.sol";
+import {IERC20} from "yield-utils-v2/contracts/token/IERC20.sol";
+import {TransferHelper} from "yield-utils-v2/contracts/token/TransferHelper.sol";
+
+/**
+@title An Fractional ERC20 wrapper/vault
+@author Sabnock01
+@notice You can use this contract to wrap and unwrap a specified ERC20 token on a fractionalized basis
+@dev The ERC20 token is set in the constructor and the wrapped tokens
+are burned upon unwrapping
+*/
+contract FractionalWrapper is ERC20 {
+    using TransferHelper for IERC20;
+    
+    ///@notice The ERC20 token to wrap
+    IERC20 public underlying;
```

Comment:
You can call it `asset`, as 4626 requires.

ERC4626 requires `asset` to be an `address`, but you can declare it as an `IERC20` and still be compliant. The ABI will still be `address` because `IERC20` is an internal type, seen only as such by the compiler.

Reply:
That makes sense.